File: nodejs/database/markets.js
import db from './db.js';
import * as utils from '../lib/utils.js';

const addMarketStatement = db.prepare(`INSERT INTO markets(system, data) VALUES (?, json(?));`);
const getMarketAtWaypointStatement = db.prepare(`SELECT data FROM markets WHERE data->>'symbol' = ?;`);
const getMarketsInSystemStatement = db.prepare(`SELECT data FROM markets WHERE system = ?;`);
const updateMarketStatement = db.prepare(`UPDATE markets SET data = json(:data) WHERE data->>'symbol' = :symbol;`);

export function getMarketAtWaypoint(symbol) {
    const data = getMarketAtWaypointStatement.get(symbol);
    if (data === undefined) {
	return null;
    }
    return JSON.parse(data.data);
}

export function getMarketsInSystem(symbol) {
    const data = getMarketsInSystemStatement.get(symbol);
    if (data === undefined) {
	return null;
    }
    return JSON.parse(data.data);
}

export function setMarket(data) {
    if (getMarketAtWaypoint(data.symbol) === null) {
	const system = utils.systemFromWaypoint(data.symbol);
	return addMarketStatement.run(system, JSON.stringify(data)).lastInsertRowid;
    }
    return updateMarketStatement.run({
	data: JSON.stringify(data),
	symbol: data.symbol,
    }).changes;
}

File: golang/pkg/database/markets.go
package database

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"time"

	"git.adyxax.org/adyxax/spacetraders/golang/pkg/model"
)

func (db *DB) LoadMarket(symbol string) (*model.Market, error) {
	var buf []byte
	if err := db.QueryRow(`SELECT data FROM markets WHERE data->>'symbol' = ?;`, symbol).Scan(&buf); err != nil {
		return nil, fmt.Errorf("failed to query row: %w", err)
	}
	var market model.Market
	if err := json.Unmarshal(buf, &market); err != nil {
		return nil, fmt.Errorf("failed to unmarshal market: %w", err)
	}
	return &market, nil
}

func (db *DB) SaveMarket(market *model.Market) error {
	data, err := json.Marshal(market)
	if err != nil {
		return fmt.Errorf("failed to marshal market: %w", err)
	}
	_, err = db.Exec(
		`INSERT INTO markets(data, updated)
                VALUES (json(:data), :updated)
         ON CONFLICT DO UPDATE SET data = :data, updated = :updated
		                       WHERE data->>'symbol' = :symbol;`,
		sql.Named("data", data),
		sql.Named("symbol", market.Symbol),
		sql.Named("updated", time.Now()),
	)
	if err != nil {
		return fmt.Errorf("failed to exec: %w", err)
	}
	return nil
}
